lib/AST/ASTWalker.cpp

```diff
@@ -425,7 +425,8 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
     if (auto def = MD->definition) {
       // Don't walk into unchecked definitions.
       if (auto expansion = dyn_cast<MacroExpansionExpr>(def)) {
-        if (!expansion->getType().isNull()) {
+        if (!expansion->getType().isNull() ||
+            Walker.shouldWalkIntoUncheckedMacroDefinitions()) {
```

Comment:
I suppose this is safer, but do you know what `expansion->getType().isNull()` was added for? We can walk over un-typechecked ASTs, so it seems like clients should be checking this instead.

Reply:
It was added in https://github.com/apple/swift/commit/aac0406. @DougGregor Do you remember why you added it?

Reply:
@DougGregor and I just chatted about it and he added this check because of a verified crash and thinks that we can fix it in a more principled way than what he did. 

I’m going to merge this as-is because it’s a low-risk change this way that we can cherry-pick to 5.9 and will provide a better fix in a follow-up PR.